File: swift/Sources/MudbaseShowcaseKanban/Models/BoardCard.swift
import Foundation

/// A single task card within a list. Mirrors `../web/src/types/card.ts`'s `BoardCard` interface.
/// `assigneeId`/`assigneeName` being `nil` is a real, distinct state from "was never set" vs.
/// "was explicitly cleared" only matters at the write boundary (see `CardsService.update`) — once
/// read back, both cases collapse to `nil` here, same as the web type's `string | null | undefined`
/// collapsing to `undefined`-shaped optionality for read purposes.
struct BoardCard: Identifiable, Equatable, Sendable {
    let id: String
    let boardId: String
    var listId: String
    var title: String
    var description: String?
    var position: Int
    var assigneeId: String?
    var assigneeName: String?
    let createdById: String
    let createdByName: String
    let createdAt: Date?

    init(document: MudbaseDocument) {
        id = document.id
        boardId = document.string("boardId") ?? ""
        listId = document.string("listId") ?? ""
        title = document.string("title") ?? ""
        description = document.string("description")
        position = document.int("position") ?? 0
        assigneeId = document.string("assigneeId")
        assigneeName = document.string("assigneeName")
        createdById = document.string("createdById") ?? ""
        createdByName = document.string("createdByName") ?? ""
        createdAt = document.createdAt
    }
}

File: mobile-flutter/lib/models/board_card.dart
/// Mirrors the web app's `BoardCard` (`web/src/types/card.ts`) and the live
/// `cards` collection schema (`boardId`, `listId`, `title`, `description?`,
/// `position`, `assigneeId?`, `assigneeName?`, `createdById`,
/// `createdByName`).
///
/// `assigneeId`/`assigneeName` being `null` is a real, distinct state (an
/// explicitly-cleared assignee - see `CardRepository.update`) from them
/// being absent on a card that never had one - both surface as `null` here
/// since Dart doesn't distinguish "key missing" from "key present with a
/// JSON `null`" once decoded into a `Map`, which is fine: this app never
/// needs to tell those two apart, only whether an assignee is currently set.
class BoardCard {
  const BoardCard({
    required this.id,
    required this.boardId,
    required this.listId,
    required this.title,
    this.description,
    required this.position,
    this.assigneeId,
    this.assigneeName,
    required this.createdById,
    required this.createdByName,
  });

  factory BoardCard.fromJson(Map<String, dynamic> json) {
    return BoardCard(
      id: json['_id'] as String? ?? '',
      boardId: json['boardId'] as String? ?? '',
      listId: json['listId'] as String? ?? '',
      title: json['title'] as String? ?? '',
      description: json['description'] as String?,
      position: (json['position'] as num?)?.toInt() ?? 0,
      assigneeId: json['assigneeId'] as String?,
      assigneeName: json['assigneeName'] as String?,
      createdById: json['createdById'] as String? ?? '',
      createdByName: json['createdByName'] as String? ?? '',
    );
  }

  final String id;
  final String boardId;
  final String listId;
  final String title;
  final String? description;
  final int position;
  final String? assigneeId;
  final String? assigneeName;
  final String createdById;
  final String createdByName;
}
